### DISCUSSION TITLE
iminuit

### DISCUSSION THREAD

**Discussion by LL-RL** [scikit-hep/iminuit#804] [Q&A]
(1) I used to call m.interactive() in jupyter lab, but it shows that 'Minuit' object has no attribute 'interactive'.
(2) When I use the following code to fit a self-defined function, the returned result is always equal to the initial value given initially, and no value is given for the fit.

```py
import numpy as np
from scipy.optimize import curve_fit
from scipy.stats import norm
from iminuit import Minuit, cost
from iminuit.cost import UnbinnedNLL, BinnedNLL

truth = 10, 10, 0.6, 10, 20, 0.8
def pdf_mulit_lognormx(x,
  A0, pos0, wid0,
  A1, pos1, wid1,
  #A2, pos2, wid2
):
return (
  #A0 * np.exp( -((np.log(x) - np.log(pos0)) / wid0)**2)
  #+ A1 * np.exp( -((np.log(x) - np.log(pos1)) / wid1)**2)
  #+ A2 * np.exp( -((np.log(x) - np.log(pos2)) / wid2)**2)
  A0*norm.cdf(x, pos0, wid0) + A1*norm.cdf(x, pos1, wid1)
)

x = np.linspace(0,1)

m = np.diff(pdf_mulit_lognormx(x, *truth))

n = np.random.default_rng(1).poisson(m)

c = cost.ExtendedBinnedNLL(n, x, pdf_mulit_lognormx)
m = Minuit(c, *truth)
m.interactive()
m.migrad() # run optimiser
m.hesse() # run covariance estimator

print(m.values) # x: 2, y: 3, z: 4
print(m.errors) # x: 1, y: 1, z: 1

# returns <ValueView A0=10.0 pos0=10.0 wid0=0.6 A1=10.0 pos1=20.0 wid1=0.8>
#           <ErrorView A0=0.0 pos0=0.0 wid0=0.0 A1=0.0 pos1=0.0 wid1=0.0>
```

**Comment by HDembinski** [comment]
Please check the version, `Minuit.interactive()` is a recent feature. Perhaps you switched environments and now use an older version?

**Reply by HDembinski** [reply]
About the second point, please check `m.fmin` to investigate the source of the problem.

**Reply by HDembinski** [reply]
You can also enable verbose output on the cost function to follow what Minuit is doing. You probably need to set limits on your parameters wid0 and wid1. However, the main issue is that your starting values pos0 and pos1 are far out of range of your x-range, which you take from 0 to 1, but pos0 and pos1 are 10 and 20.

**Reply by LL-RL** [reply]
Dear HD embinski,
Thank you for your help, the above problem has been solved with your help, but today I have the following problem.
When I use the following code to fit the model parameters, there is a big, big fitting result. Why is this?

import numpy as np
from scipy.optimize import curve_fit
from scipy.stats import norm, lognorm
from iminuit import Minuit, cost
from iminuit.cost import UnbinnedNLL, BinnedNLL
rng = np.random.default_rng(1)
#truth = 6.43376888, 10.98791702,  0.90578677, 16.3668948,   8.95121026,  0.21826914
def pdf_mulit_lognormx(x, 
                       mu0, A0, sigma0, 
                       mu1, A1, sigma1, 
                       #A2, pos2, wid2
                      ):
    return (
             np.cumsum(A0 * np.exp( -((np.log(x) - np.log(mu0)) / sigma0)**2) 
            + A1 * np.exp( -((np.log(x) - np.log(mu1)) / sigma1)**2))
            #+ A2 * np.exp( -((np.log(x) - np.log(pos2)) / wid2)**2)
        #A0*lognorm.pdf(x, mu0, sigma0) + A1*lognorm.pdf(x, mu1, sigma1)
           )

eps = np.finfo(np.float64).eps
hist, edges = np.histogram(data1, bins = 20, density = True, range = [min(data_source['flux']), max(data_source['flux'])])
#width = np.diff(edges)[0]
#print(width)
nll = BinnedNLL(hist, edges, pdf_mulit_lognormx)
print(nll)

minimize_binned = Minuit(nll, mu0 = 16.69, A0 = 0.05, sigma0=0.221, mu1=6.719, A1=0.064, sigma1=0.892)
minimize_binned.limits['A0', 'A1'] = (eps, None)
minimize_binned.limits['sigma0','sigma1'] = (eps, None)
minimize_binned.limits['mu0', 'mu1'] = (0, None)
#minimize_binned.interactive()
minimize_binned.migrad()

return
mu0 | 1.6568e9 | 0.0000e9
A0 | 7.5579e6 | 0.0000e6
sigma0 | 5.9602e8 | 0.0000e8 |  
mu1 | 1.64e9 | 0.00e9 | 
A1 | 7.3772e6 | 0.0000e6 |  
sigma1 | 2.2423e8 | 0.0000e8

**Reply by HDembinski** [reply]
You don't have to use eps in limits, the limit is an exclusive interval.

I don't see what the error is, you have to look yourself. Also, please use markup when you post code, to make it readable. I added markup to your original post.
